app/src/main/java/com/rockthevote/grommet/ui/eventFlow/PartnerLoginViewModel.kt

```diff
@@ -0,0 +1,115 @@
+package com.rockthevote.grommet.ui.eventFlow
+
+import androidx.lifecycle.*
+import com.rockthevote.grommet.data.api.RockyService
+import com.rockthevote.grommet.data.db.dao.PartnerInfoDao
+import com.rockthevote.grommet.data.db.model.PartnerInfo
+import com.rockthevote.grommet.util.coroutines.DispatcherProvider
+import com.rockthevote.grommet.util.coroutines.DispatcherProviderImpl
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+/**
+ * Created by Mechanical Man on 5/16/20.
+ */
+class PartnerLoginViewModel(
+        dispatchers: DispatcherProvider = DispatcherProviderImpl(),
+        private val rockyService: RockyService,
+        private val partnerInfoDao: PartnerInfoDao
+) : ViewModel() {
+
+    val partnerInfoId: LiveData<Long> =
+            Transformations.map(partnerInfoDao.getCurrentPartnerInfo()) { result ->
+                result.partnerInfoId
+            }
+
+    private val _partnerLoginState = MutableLiveData<PartnerLoginState>(PartnerLoginState.Init)
+    val partnerLoginState: LiveData<PartnerLoginState> = _partnerLoginState
+
+    private val _effect = MutableLiveData<PartnerLoginState.Effect?>()
+    val effect: LiveData<PartnerLoginState.Effect?> = _effect
+
+    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+        // TODO Should we handle this another way? Is it needed?
+        Timber.e(throwable)
+    }
+
+    private val supervisorJob = SupervisorJob()
```

Comment:
Two things on use of SupervisorJob.

1. You don't really need it for this scenario. You can, but it's overkill. SupervisorJob allows for running multiple parallel tasks, if one of those tasks unexpectedly trip an exception, it won't bring down all the others. Using `viewModelScope` (an extension function of ViewModel) should be enough because you're only making one call. I only used SupervisorJob in the other VM because it's making a large amount of call simultaneously. `viewModelScope` also cleans itself up without you having to think about it.

2. If you do use `SupervisorJob` you need to cancel all its children in the `onCleared` method of the ViewModel so it doesn't leak.

For `viewModelScope`, you should be able to just call `viewModelScope.launch(dispatchers.io)` to launch something in the IO scope with the DispatcherProvider https://developer.android.com/topic/libraries/architecture/coroutines

Reply:
Ah, just realized you're using `viewModelScope` in `init`, it's about that simple.

Reply:
thanks, not gonna lie, lotta copy/pasta coding w/ the coroutines, starting to get it but def not as knowledgeable as you

Reply:
So what's the correct way to scope the coroutine in my API call, 

`viewModelScope.launch(dispatchers.io)` or `rockyRequestScope.launch ` ?